File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        Map<Character, Integer> need = new HashMap<>();
        Map<Character, Integer> window = new HashMap<>();
        int formed = 0;
        int left = 0, right = 0;
        int minLen = Integer.MAX_VALUE;
        int minLeft = 0;

        for (char c : t.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        int required = need.size();

        while (right < s.length()) {
            char c = s.charAt(right);
            window.put(c, window.getOrDefault(c, 0) + 1);
            
            if (need.containsKey(c) && window.get(c).equals(need.get(c))) {
                formed++;
            }

            while (formed == required) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minLeft = left;
                }

                char leftChar = s.charAt(left);
                window.put(leftChar, window.get(leftChar) - 1);

                if (need.containsKey(leftChar) && window.get(leftChar) < need.get(leftChar)) {
                    formed--;
                }

                left++;
            }

            right++;
        }

        return minLen == Integer.MAX_VALUE ? "" : s.substring(minLeft, minLeft + minLen);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.kt
class Solution {
    fun minWindow(s: String, t: String): String {
        val need = mutableMapOf<Char, Int>()
        val window = mutableMapOf<Char, Int>()

        for (c in t) {
            need[c] = need.getOrDefault(c, 0) + 1
        }

        var formed = 0
        val required = need.size
        var left = 0
        var right = 0
        var minLen = Int.MAX_VALUE
        var minLeft = 0

        while (right < s.length) {
            val c = s[right]
            window[c] = window.getOrDefault(c, 0) + 1

            if (need.containsKey(c) && window[c] == need[c]) {
                formed++
            }

            while (formed == required) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1
                    minLeft = left
                }

                val leftChar = s[left]
                window[leftChar] = window.getOrDefault(leftChar, 0) - 1

                if (need.containsKey(leftChar) && window[leftChar]!! < need[leftChar]!!) {
                    formed--
                }

                left++
            }

            right++
        }

        return if (minLen == Int.MAX_VALUE) "" else s.substring(minLeft, minLeft + minLen)
    }
}